### PR TITLE
observability: implement server interceptor for logging

### PR DESCRIPTION
**Pull request by DNVindhya**
This PR implements Server Interceptor for logging i.e InternalLoggingServerInterceptor. Events are intercepted and a log record is written to logging as part of intercepting methods.

RPC messages will be logged if they are of following types:
- `com.google.protobuf.Message`
- `byte[]`

### CODE REVIEW

**Comment by DNVindhya** [comment]
CC @sanjaypujare

File: observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java [hunk 1]
```diff
@@ -16,32 +16,224 @@
...
   }
 
+  private InternalLoggingServerInterceptor(LogHelper helper) {
+    this.helper = helper;
+  }
+
   @Override
   public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
       Metadata headers, ServerCallHandler<ReqT, RespT> next) {
     // TODO(dnvindhya) implement the interceptor
```
Comment (sanjaypujare) [inline observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java:89]: remove the comment

File: observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java [hunk 2]
```diff
@@ -16,32 +16,224 @@
...
+      // Expected exceptions to be caught:
+      // 1. IllegalArgumentException
+      // 2. NullPointerException
+      logger.log(Level.SEVERE, "Unable to log request header", e);
+    }
+
+    ServerCall<ReqT, RespT> wrapperCall =
+        new SimpleForwardingServerCall<ReqT, RespT>(call) {
+          @Override
+          public void sendMessage(RespT message) {
```
Comment (sanjaypujare) [inline observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java:131]: nit: if the lexical order of methods also reflects the order in which these events take place it will be good. So how about `sendHeaders` and then `sendMessage` ?

File: observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java [hunk 3]
```diff
@@ -16,32 +16,224 @@
...
+              helper.logTrailer(
+                  seq.getAndIncrement(),
+                  serviceName,
+                  methodName,
+                  status,
+                  trailers,
+                  EventLogger.LOGGER_SERVER,
+                  rpcId,
+                  null);
+            } catch (Exception e) {
```
Comment (sanjaypujare) [inline observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java:179]: I wonder if putting this catch clause and the `logger.log` statement should be moved into the corresponding `helper.log*` methods. This will make the caller side code simpler. This needs to be done for both client and server so we can do this in another PR

File: observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java [hunk 4]
```diff
@@ -225,6 +236,12 @@ void logTrailer(
         .setEventLogger(eventLogger)
         .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
         .setRpcId(rpcId);
+    if ((pair != null) && (pair.size != 0)) {
```
Comment (sanjaypujare) [inline observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java:239]: nit: the parens around `pair != null` etc are not required.

File: observability/src/test/java/io/grpc/observability/LoggingTestHelper.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import io.grpc.testing.protobuf.SimpleServiceGrpc;
+import org.mockito.ArgumentMatchers;
+
+public class LoggingTestHelper {
+
+  static ServerCallHandler<String, Integer> anyCallHandler() {
+    return ArgumentMatchers.any();
+  }
+
+  static String unaryRpc(
```
Comment (sanjaypujare) [inline observability/src/test/java/io/grpc/observability/LoggingTestHelper.java:32]: This static method called `unaryRpc` is a bit misleading. What it does is to make a unary call thru a client stub. How about naming it as `makeUnaryRpcViaClientStub` or something like that?

File: observability/src/test/java/io/grpc/observability/LoggingTestHelper.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import io.grpc.ServerCallHandler;
+import io.grpc.stub.StreamObserver;
+import io.grpc.testing.protobuf.SimpleRequest;
+import io.grpc.testing.protobuf.SimpleResponse;
+import io.grpc.testing.protobuf.SimpleServiceGrpc;
+import org.mockito.ArgumentMatchers;
+
+public class LoggingTestHelper {
+
+  static ServerCallHandler<String, Integer> anyCallHandler() {
```
Comment (sanjaypujare) [inline observability/src/test/java/io/grpc/observability/LoggingTestHelper.java:28]: Looks like this is dead code. Where is `anyCallHandler` used?

File: observability/src/test/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptorTest.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+          ServerCall<byte[], byte[]> call,
+          Metadata headers) {
+        return new ServerCall.Listener<byte[]>() {};
+      }
+    };
+
+    // We expect the contents of the "grpc-timeout" header to be installed the context
+    Context.current()
+        .withDeadlineAfter(1, TimeUnit.SECONDS, Executors.newSingleThreadScheduledExecutor())
+        .run(new Runnable() {
```
Comment (sanjaypujare) [inline observability/src/test/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptorTest.java:276]: Consider using lambda. 

```
() -> {
          ServerCall.Listener<byte[]> unused =
              factory.create()
                  .interceptCall(noopServerCall, new Metadata(), noopHandler);
        } 
```

File: observability/src/test/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptorTest.java [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+                        .build();
+                  }
+
+                  @Override
+                  public String getAuthority() {
+                    return "the-authority";
+                  }
+                },
+                clientInitial,
+                new ServerCallHandler<byte[], byte[]>() {
```
Comment (sanjaypujare) [inline observability/src/test/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptorTest.java:146]: Consider using lambda:

```
                    (call, headers) -> {
                      interceptedLoggingCall.set(call);
                      return mockListener;
                    }
```

**Review by sanjaypujare — COMMENTED** [review]
Looks good. Will take a look again after changes

File: observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java (hunk 1 above)
Comment (DNVindhya) [reply observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java:89]: Done

File: observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java (hunk 2 above)
Comment (DNVindhya) [reply observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java:131]: Updated

File: observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java (hunk 4 above)
Comment (DNVindhya) [reply observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java:239]: Removed extra parentheses.

File: observability/src/test/java/io/grpc/observability/LoggingTestHelper.java (hunk 6 above)
Comment (DNVindhya) [reply observability/src/test/java/io/grpc/observability/LoggingTestHelper.java:28]: It was indeed dead code, removed it.

File: observability/src/test/java/io/grpc/observability/LoggingTestHelper.java (hunk 5 above)
Comment (DNVindhya) [reply observability/src/test/java/io/grpc/observability/LoggingTestHelper.java:32]: Renamed `unaryRpc` to `makeUnaryRpcViaClientStub`

File: observability/src/test/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptorTest.java (hunk 8 above)
Comment (DNVindhya) [reply observability/src/test/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptorTest.java:146]: Done

File: observability/src/test/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptorTest.java (hunk 7 above)
Comment (DNVindhya) [reply observability/src/test/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptorTest.java:276]: Done

**Review by sanjaypujare — APPROVED** [review]
LGTM
